tests/bit_fade.c

```diff
@@ -58,55 +60,52 @@ static int pattern_fill(int my_cpu, testword_t pattern)
                 continue;
             }
             test_addr[my_cpu] = (uintptr_t)p;
-            do {
-                write_word(p, pattern);
-            } while (p++ < pe); // test before increment in case pointer overflows
-            do_tick(my_cpu);
-            BAILOUT;
-        } while (!at_end && ++pe); // advance pe to next start point
-    }
-
-    flush_caches(my_cpu);
-
-    return ticks;
-}
-
-static int pattern_check(int my_cpu, testword_t pattern)
-{
-    int ticks = 0;
-
-    for (int i = 0; i < vm_map_size; i++) {
-        testword_t *start = vm_map[i].start;
-        testword_t *end   = vm_map[i].end;
-
-        testword_t *p  = start;
-        testword_t *pe = start;
-
-        bool at_end = false;
-        do {
-            // take care to avoid pointer overflow
-            if ((end - pe) >= SPIN_SIZE) {
-                pe += SPIN_SIZE - 1;
-            } else {
-                at_end = true;
-                pe = end;
+            if (!check) {
+#if HAND_OPTIMISED
+#ifdef __x86_64__
+                uint64_t length = pe - p + 1;
+                __asm__  __volatile__ ("\t"
+                    "rep    \n\t"
+                    "stosq  \n\t"
+                    :
+                    : "c" (length), "D" (p), "a" (pattern)
+                    :
+                );
+                p = pe;
```

Comment:
Looking more carefully, I think this should be 'p = pe + 1' or 'p += length'. It shouldn't cause any problems, as it just means we rewrite the last location of each spin at the start of the next spin. My fault, as the code is the same in mov_inv_fixed.c

Reply:
I have changed both p reassignments in both tests.

Reply:
The `stage %= 3;` statement in my patch clamps stage to 0-2, but I don't see that having an impact on the ` if (stage != last_stage) {` test in the switch after the modulo 3.
At the moment, I can't see how that test condition could ever be false, due to the `last_stage = stage;` assignment after the switch. IIUC, when the program reaches cases 1 (or 4) for `stage`, `last_stage` was set to 0 or 3 by that assignment during the previous run of that function, so `last_stage` and `stage` compare unequal (before and after my patch). However, it's getting late, so I'll re-examine the code later :)